### PR TITLE
babe: support online configuration upgrades

### PR DESCRIPTION
**Pull request by sorpaas**
This is an attempt to support online configuration upgrades for BABE engine. Still needs to write the migration scripts, but would love to get some reviews.

Currently, BABE's configuration is only read at genesis block. It's then fixed, and cannot be changed. As a result, it's not possible to perform online upgrades, for example, from primary-slots only to primary-and-secondary-slots. This PR tries to fix that, by doing the following:

* Split the configuration of BABE into genesis configuration and epoch configuration. 
* Genesis configuration is fixed and only read at genesis. Epoch configuration is read at every epoch, and stored alongside the epoch.
* Genesis configuration contains information that cannot be changed. This includes the slot duration, genesis authorities and genesis randomness.
* Epoch configuration contains other information that can be changed. This includes the value of `c`, and whether the engine supports secondary-slots. The encoding of epoch configuration can also be flexibly extended in the future.
* At every next epoch descriptor, the *parent block*'s `epoch_configuration` is read, and used as the epoch configuration for the epoch described by the next epoch descriptor. This config is then used by authorship and verification rules.

### CODE REVIEW

**Comment by rphmeier** [comment]
It will need light client support, I think, because any changes in configuration is technically a hard-fork and requires finality to process.

**Comment by sorpaas** [comment]
Discussed with Rob. I'm going to add the config struct also to `NextEpochDescriptor`, to fix the light client issue.

* For full node, the engine reads `BabeEpochConfiguration` from runtime, and check that it matches the data provided in `NextEpochDescriptor`.
* For light client, the engine just accepts whatever `BabeEpochConfiguration` provided in `NextEpochDescriptor`.

**Comment by sorpaas** [comment]
Changed this a little bit, and I think I've got a design that I'm satisfied with a) maintains backward compatibility, b) support light clients, and c) extensible:

* The genesis config `BabeGenesisConfiguration` is unchanged (same as old `BabeConfiguration`), and the runtime interface is unchanged.
* The epoch config now just contains two values that I think is safe to change. `c` and `secondary_slots`. Not including epoch length as I'm just not quite sure. It can be added in the future.
* A new digest is added, `NextConfigDescriptor`. The digest is optional, and if it is there together with `NextEpochDescriptor`, then the epoch after the current epoch will use the configuration as described by the descriptor. This is backward compatible of current BABE because blocks without `NextConfigDescriptor` is valid.
* In the future, when we need to add new values to epoch config, it can just be defined as an additional digest type, and enact separately with the same rule. Changing genesis config will still require runtime interface change, though (so to support VRF from genesis in #5501 the runtime version there sill have to be changed).

File: client/consensus/babe/src/lib.rs [hunk 1]
```diff
@@ -995,11 +1046,15 @@ impl<Block, Client, Inner> BlockImport<Block> for BabeBlockImport<Block, Client,
 
 			let viable_epoch = epoch_changes.viable_epoch(
 				&epoch_descriptor,
-				|slot| self.config.genesis_epoch(slot),
+				|slot| Epoch::genesis(&self.config, slot)
 			).ok_or_else(|| {
 				ConsensusError::ClientImport(Error::<Block>::FetchEpoch(parent_hash).into())
 			})?;
 
+      let epoch_config = next_config_digest.unwrap_or_else(
```
Comment (andresilva) [inline client/consensus/babe/src/lib.rs:1054]:
```suggestion
			let epoch_config = next_config_digest.unwrap_or_else(
```

**Comment by andresilva** [comment]
In a follow-up PR we should add a method to the babe pallet, callable by root, that allows signaling a config change. This should queue the change so that the event is sent together with the next epoch descriptor.

File: client/consensus/babe/src/lib.rs (hunk 1 above)
Comment (sorpaas) [reply client/consensus/babe/src/lib.rs:1054]: Ah! Not sure why my editorconfig suddenly failed here.

**Comment by sorpaas** [comment]
NOTE: This needs to be merged together with #5501 to avoid multiple migrations. I'll do Kusama testing tomorrow.

**Comment by sorpaas** [comment]
Finished test syncing Kusama in #5501.

**Pull request merged**